nemo_aligner/data/nlp/builders.py

```diff
@@ -52,15 +54,70 @@
 from nemo_aligner.utils.utils import collate_with_batch_max_sequence_length
 
 
-def build_dataset_generic(cls, cfg, data_prefix, data_impl, num_samples, seq_length, seed, tokenizer, name):
+class ChunkedJsonl:
+    def __init__(self, path_placeholder, n_chunks, n_examples_per_chunk):
+        self.CHUNK_ID_STRING = "CHUNK_ID"
+        assert (
+            self.CHUNK_ID_STRING in path_placeholder
+        ), f"path_placehold ({path_placeholder}) does not have the CHUNK_ID_STRING ({self.CHUNK_ID_STRING})"
+        self.path_placeholder = path_placeholder
+
+        # get the maximum number of chunks
+        max_n_chunks = 0
+        for i in range(n_chunks):
+            max_n_chunks = i + 1  ## fix zero-indexing
+            if not os.path.exists(self.path_placeholder.replace(self.CHUNK_ID_STRING, str(i))):
+                break
+        assert max_n_chunks > 0, f"no files match the required path {path_placeholder}"
+        self.n_chunks = min(n_chunks, max_n_chunks)
+
+        print(f"Initializing chunked jsonl...")
+        lengths = [n_examples_per_chunk for _ in range(self.n_chunks)]
+        print(f"Number of Chunks = {self.n_chunks} | Number of Examples = {n_examples_per_chunk * self.n_chunks}")
+        self._lengths = np.asarray(lengths)
+        self._length_accumulated = np.cumsum(lengths)
+
+    def __len__(self):
+        return self._lengths.sum()
+
+    def __getitem__(self, i):
+        if i >= len(self):
+            raise ValueError(f"The item idx {i} is greater than the length of the dataset ({len(self)})")
+        chunk_id = np.searchsorted(self._length_accumulated, i, side="right")
+        idx_in_chunk = i if chunk_id == 0 else i - self._length_accumulated[chunk_id - 1]
+        with open(self.path_placeholder.replace(self.CHUNK_ID_STRING, str(chunk_id))) as f:
+            for line_idx, l in enumerate(f):
+                if line_idx == idx_in_chunk:
+                    return json.loads(l)
```

Comment:
I believe we do need random access if we want to shuffle the data, but this is a good point. Can this be addressed in a subsequent PR since it might require some thought?

Reply:
discussed offline to defer this to a followup PR